source/brailleDisplayDrivers/braillePen.py

```diff
@@ -0,0 +1,143 @@
+			try:
+				inputCore.manager.executeGesture(InputGesture(braille, command))
+			except inputCore.NoInputGestureAction:
+				pass
+	
+	def display(self, cells):
+		pass
+
+	gestureMap = inputCore.GlobalGestureMap({
+		"globalCommands.GlobalCommands": {
+			"kb:upArrow": ("br(braillePen):space+dot1",),
+			"kb:downArrow": ("br(braillePen):space+dot4",),
+			"kb:leftArrow": ("br(braillePen):space+dot3",),
+			"kb:rightArrow": ("br(braillePen):space+dot6",),
+			"kb:enter": ("br(braillePen):space+dot8"),			
+			"kb:backspace": ("br(braillePen):space+dot7","br(braillePen):space+dot1+dot2"),
+			"kb:delete": ("br(braillePen):space+dot1+dot4+dot5",),
+			"kb:windows": ("br(braillePen):space+dot3+dot4",),
+			"kb:alt": ("br(braillePen):space+dot1+dot3+dot4",),
+			"kb:escape": ("br(braillePen):space+dot1+dot5",),
+			"kb:shift+tab": ("br(braillePen):space+dot2+dot3+dot4+dot5+dot8",),
+			"kb:tab": ("br(braillePen):space+dot2+dot3+dot4+dot5+dot7",),			
+			"toggleInputHelp": ("br(braillePen):space+dot3+dot6+dot7",),
+			"braille_toggleTether": ("br(braillePen):space+dot1+dot2+dot6",),
+			"kb:home": ("br(braillePen):space+dot2+dot3",),
+			"kb:end": ("br(braillePen):space+dot5+dot6",),
+			"sayAll": ("br(braillePen):space+dot1+dot2+dot4+dot5+dot6",),
+			"dateTime": ("br(braillePen):space+dot2+dot3+dot4",),
+			"reportCurrentSelection": ("br(braillePen):space+dot1+dot2+dot4+dot5+dot6+dot7",),
+			"showGui": ("br(braillePen):space+dot1+dot2+dot4+dot5",),
+			"kb:alt+tab": ("br(braillePen):space+dot1+dot3+dot5",),
+			"say_battery_status": ("br(braillePen):space+dot1+dot2+dot7",),
+			"title": ("br(braillePen):space+dot2+dot3+dot4+dot5",),
+			"reportCurrentFocus": ("br(braillePen):space+dot1+dot2+dot4+dot7",),
+			"reportStatusLine": ("br(braillePen):space+dot2+dot3+dot4+dot7",),
+			"speakForeground": ("br(braillePen):space+dot1+dot2+dot4+dot5+dot7",),
+			"kb:control+leftArrow": ("br(braillePen):space+dot2",),
+			"kb:control+rightArrow": ("br(braillePen):space+dot5",),
+			"kb:control+home": ("br(braillePen):space+dot2+dot3+dot7",),
+			"kb:control+end": ("br(braillePen):space+dot5+dot6+dot7",),
+		},
+	})
+
+class InputGesture(braille.BrailleDisplayGesture, brailleInput.BrailleInputGesture):
+
+	source = BrailleDisplayDriver.name
+
+	def __init__(self, brailleKeys,controlKeys):
+		super(InputGesture, self).__init__()
+		spaceKey=False
+		brailleKeysExtended = ord(brailleKeys)
+		controlKeysExtended = ord(controlKeys)
+		if controlKeysExtended & 1 >0:
+			spaceKey=True
+		if controlKeysExtended & 2 >0:
+			brailleKeysExtended = brailleKeysExtended | 0x40
+		if controlKeysExtended & 4 >0:
+			brailleKeysExtended = brailleKeysExtended | 0x80
+		brailleKeysList=[BP12_BRAILLE_KEYS[num] for num in xrange(8) if (brailleKeysExtended>>num)&1]
+		spaceKeyList = ["space"]
+		if spaceKey:
+			selectedKeys=spaceKeyList+brailleKeysList
+		else:
+			selectedKeys=brailleKeysList
+		self.id="+".join(set(selectedKeys))
+		self.keys = set(selectedKeys)
```

Comment:
There seem to be a few unnecessary lists/sets being created here. I think you could cut this down to something like this:

```
keys = self.keys = {BP12_BRAILLE_KEYS[num] for num in xrange(8) if (brailleKeysExtended >> num) & 1}
if spaceKey:
	keys.add("space")
self.id = "+".join(keys)
```